modules/renter/hostdb/upload.go

```diff
@@ -201,17 +202,36 @@ outer:
 		}
 	}
 
-	// form new contracts from randomly-picked nodes
+	// Also exclude hosts on the pool's blacklist, and hosts we're already
+	// connected to.
+	exclude = append(exclude, p.blacklist...)
+	for _, h := range p.hosts {
+		exclude = append(exclude, h.Address())
+	}
+
+	// Ask the hostdb for random hosts. We always ask for at least 10, to
+	// avoid selecting the same uncooperative hosts over and over.
+	ask := n
+	if ask < 10 {
+		ask = 10
+	}
 	p.hdb.mu.Lock()
-	randHosts := p.hdb.randomHosts(n*2, exclude)
+	randHosts := p.hdb.randomHosts(ask, exclude)
 	p.hdb.mu.Unlock()
+
+	// Form new contracts with the randomly-picked hosts. If a contract can't
+	// be formed, add the host to the pool's blacklist.
 	for _, host := range randHosts {
 		contract, err := p.hdb.newContract(host, p.filesize, p.duration)
 		if err != nil {
+			p.hdb.log.Printf("couldn't form contract with %v: %v", host.NetAddress, err)
+			p.blacklist = append(p.blacklist, host.NetAddress)
 			continue
 		}
 		hu, err := p.hdb.newHostUploader(contract)
 		if err != nil {
+			p.hdb.log.Printf("couldn't create uploader for %v: %v", host.NetAddress, err)
```

Comment:
a log here is probably not needed, unless you are unable to make uploaders with anybody or some fractional amount of them.